File: src/main/scala/org/unipi/matrixnorm/MapperValue.scala
package org.unipi.matrixnorm

import java.io.{DataInput, DataOutput, IOException}
import org.apache.hadoop.io.Writable

class MapperValue(var matrixIndex: Integer, var rowIndex: Integer, var colValue: Double) extends Writable {

  def this() = this(matrixIndex = 0, rowIndex = 0, colValue = 0.0)

  @throws[IOException]
  override def readFields(in: DataInput): Unit = {
    val arr = in.readLine().split("\t").map(_.trim)
    this.matrixIndex = BigInt(arr(0)).intValue()
    this.rowIndex = BigInt(arr(1)).intValue()
    this.colValue = BigDecimal(arr(2)).doubleValue()
  }

  @throws[IOException]
  override def write(out: DataOutput): Unit = {
    val data = Array(this.matrixIndex.toString, this.rowIndex.toString, this.colValue.toString).mkString("\t")
    out.writeBytes(data)
  }

  def canEqual(a: Any): Boolean = a.isInstanceOf[MapperValue]

  override def equals(that: Any): Boolean =
    that match {
      case that: MapperValue => that.canEqual(this) && this.hashCode == that.hashCode
      case _ => false
    }

  override def hashCode: Int = {
    val prime = 37
    var result = 1
    result = prime * result + matrixIndex
    result = prime * result + rowIndex
    result = prime * result + colValue.toInt
    result
  }
}

File: src/main/java/org/unipi/matrixnorm/MapperValue.java
package org.unipi.matrixnorm;

import org.apache.hadoop.io.Writable;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

import java.util.Arrays;

public class MapperValue implements Writable {

    int rowIndex    = 0;
    double colValue = 0;

    public MapperValue(){}

    public MapperValue(int rowIndex, double colValue) {
        this.rowIndex = rowIndex;
        this.colValue = colValue;
    }

    public void readFields(DataInput in) throws IOException {
        rowIndex = in.readInt();
        colValue = in.readDouble();
    }

    public void write(DataOutput out) throws IOException {
        out.writeInt(rowIndex);
        out.writeDouble(colValue);
    }

    public boolean equals(Object o) {
        if (o == null || !this.getClass().equals(o.getClass())) {
            return false;
        }
        MapperValue other = (MapperValue) o;
        return this.hashCode() == other.hashCode();
    }

    public int hashCode() {
        int prime = 37;
        int result = 1;
        result = prime * result + rowIndex;

        long l = Double.doubleToLongBits(colValue);

        result = prime * result + (int)(l ^ (l >>> 32));
        return result;
    }
}
